pam/internal/adapter/nativemodel.go

```diff
@@ -0,0 +1,633 @@
+			})
+		}
+
+		if len(m.authModes) == 1 {
+			return m, sendEvent(authModeSelected{id: m.authModes[0].Id})
+		}
+
+		var choices choicesList
+		for _, am := range m.authModes {
+			choices = append(choices, choicePair{id: am.Id, label: am.Label})
+		}
+
+		id, err := m.promptForChoice("Authentication mode selection", choices,
+			"Select authentication mode")
+		if errors.Is(err, errGoBack) {
+			m.authModes = nil
+			if len(m.availableBrokers) < 2 {
+				return m, tea.Sequence(sendEvent(ChangeStage{proto.Stage_userSelection}),
+					sendEvent(nativeUserSelection{}))
+			}
+			return m, tea.Sequence(sendEvent(ChangeStage{proto.Stage_brokerSelection}),
+				sendEvent(nativeBrokerSelection{}))
+		}
+		if err != nil {
+			return m, sendEvent(pamError{
+				status: pam.ErrSystem,
+				msg:    fmt.Sprintf("broker selection error: %v", err),
+			})
+		}
+
+		return m, sendEvent(authModeSelected{id: id})
+
+	case authModeSelected:
+		m.selectedAuthMode = msg.id
+
+	case UILayoutReceived:
+		m.uiLayout = msg.layout
+
+	case startAuthentication:
+		if m.uiLayout == nil {
+			return m, sendEvent(pamError{
+				status: pam.ErrSystem,
+				msg:    "Can't authenticate without an UI layout selected",
+			})
+		}
+		return m, sendEvent(nativeChallengeRequested{})
+
+	case nativeChallengeRequested:
+		return m, m.startChallenge()
+
+	case isAuthenticatedResultReceived:
+		access := msg.access
+		authMsg, err := dataToMsg(msg.msg)
+		if err != nil {
+			return m, sendEvent(pamError{status: pam.ErrSystem, msg: err.Error()})
+		}
+
+		switch access {
+		case brokers.AuthGranted:
+			return m, sendPamError(m.sendInfo(authMsg))
+		case brokers.AuthNext:
+			m.uiLayout = nil
+			return m, sendPamError(m.sendInfo(authMsg))
+		case brokers.AuthDenied, brokers.AuthRetry:
+			return m, sendPamError(m.sendError(authMsg))
+		case brokers.AuthCancelled:
+			return m, sendEvent(isAuthenticatedCancelled{})
+		default:
+			return m, sendPamError(m.sendError("Access %q is not valid", access))
+		}
+
+	case isAuthenticatedCancelled:
+		m.uiLayout = nil
+		// FIXME: Should we go back to previous stage too?
+	}
+
+	return m, nil
+}
+
+func checkForPromptReplyValidity(reply string) error {
+	switch reply {
+	case nativeCancelKey:
+		return errGoBack
+	case "", "\n":
+		return errEmptyResponse
+	default:
+		return nil
+	}
+}
+
+func (m nativeModel) promptForInput(style pam.Style, prompt string) (string, error) {
+	resp, err := m.pamMTx.StartStringConvf(style, "%s: ", prompt)
+	if err != nil {
+		return "", err
+	}
+	return resp.Response(), checkForPromptReplyValidity(resp.Response())
+}
+
+func (m nativeModel) promptForNumericInput(style pam.Style, prompt string) (int, error) {
+	for {
+		out, err := m.promptForInput(style, prompt)
+		if err != nil {
+			return -1, err
+		}
+		intOut, err := strconv.Atoi(out)
+		if err == nil {
+			return intOut, nil
+		}
+
+		err = m.sendError("Provided input can't be parsed as integer value")
+		if err != nil {
+			return -1, err
+		}
```

Comment:
So, the logic here was to retry in case of parsing failures, and just just notify an try again in case that happens.

Maybe it can be handled in the caller too, but I didn't want to repeat the same logic...

Reply:
Yeah, I'm not 100% sure on what could be the right behavior here also. I don't like having this kind of loops on functions that don't indicate them straight away. When I look at `promptForNumericInput`, I expect a single prompt and then, if something goes wrong, the caller handles it. Maybe we should move the loop logic to the "main" function?

Reply:
I've changed it to use another function that recurses if needed, to avoid having to re-handle the same error in various places.